### PR TITLE
CWMSVue 594 - Added Category and Group office IDs to TimeSeriesGroup and LocationGroup DAOs

### PR DESCRIPTION
**Pull request by zack-rma**
CWMSVue 594 - Added Category and Group office IDs to TimeSeriesGroup and LocationGroup DAOs

### CODE REVIEW

File: cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java [hunk 1]
```diff
@@ -135,15 +135,20 @@ Boolean.class, false, metrics, name(LocationGroupController.class.getName(),
...
-                    + "the location_group whose data is to be included in the response")
+                + "the location_group whose data is to be included in the response")
         },
         queryParams = {
             @OpenApiParam(name = OFFICE, required = true, description = "Specifies the "
-                    + "owning office of the location group whose data is to be included "
-                    + "in the response."),
+                + "owning office of the location group whose data is to be included "
+                + "in the response."),
+            @OpenApiParam(name = GROUP_OFFICE_ID, required = true, description = "Specifies the "
```
Comment (zack-rma) [inline cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java:144]: I could use some input on the language for describing the different office parameters.

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java [hunk 2]
```diff
@@ -82,6 +82,10 @@ private Timer.Context markAndTime(String subject) {
                         + " the assigned timeseries in the returned timeseries groups. (default: true)"),
                 @OpenApiParam(name = TIMESERIES_CATEGORY_LIKE, description = "Posix <a href=\"regexp.html\">regular expression</a> "
                         + "matching against the timeseries category id"),
+                @OpenApiParam(name = CATEGORY_OFFICE_ID, description = "Specifies the owning office of the "
```
Comment (zack-rma) [inline cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java:106]: I could use some input on the language for describing the different office parameters.

**Review by zack-rma — COMMENTED** [review]
Looking for input on the wording of the documentation for the new office id parameters

File: cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java (hunk 1 above)
Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java:144]: That looks fine to me

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java (hunk 2 above)
Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java:106]: looks good to me

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java [hunk 3]
```diff
@@ -220,6 +233,16 @@ public void create(@NotNull Context ctx) {
             String body = ctx.body();
             ContentType contentType = Formats.parseHeader(formatHeader, TimeSeriesGroup.class);
             TimeSeriesGroup deserialize = Formats.parseContent(contentType, body, TimeSeriesGroup.class);
+
+            if (!deserialize.getTimeSeriesCategory().getOfficeId().equalsIgnoreCase(CWMS_OFFICE)
+                    && (!deserialize.getOfficeId().equalsIgnoreCase(deserialize.getTimeSeriesCategory().getOfficeId())
+                    || deserialize.getOfficeId().equalsIgnoreCase(CWMS_OFFICE))) {
+                CdaError re = new CdaError("Office ID cannot be CWMS and must match the TimeSeries category office ID");
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java]: Throw an IllegalArgumentException instead with the error message, the error code will then get handled and serialized correctly (with incident identifier, error code 400 etc)

Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java]: Should say the "Time Series Group office id cannot be ...."

File: cwms-data-api/src/test/java/cwms/cda/api/LocationCategoryControllerTestIT.java [hunk 4]
```diff
@@ -119,4 +119,178 @@ void test_create_read_delete() throws Exception {
...
+		.when()
+			.redirects().follow(true)
+			.redirects().max(3)
+			.post("/location/category")
+		.then()
+			.log().ifValidationFails(LogDetail.ALL,true)
+		.assertThat()
+			.statusCode(is(HttpServletResponse.SC_CONFLICT));
+		//Read Empty
+		given()
```
Comment (adamkorynta) [inline cwms-data-api/src/test/java/cwms/cda/api/LocationCategoryControllerTestIT.java]: Don't need a subsequent read test, the previous SC_CONFLICT is enough

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java (hunk 3 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java]: Updated to throw IllegalArgumentException

File: cwms-data-api/src/test/java/cwms/cda/api/LocationCategoryControllerTestIT.java (hunk 4 above)
Comment (zack-rma) [reply cwms-data-api/src/test/java/cwms/cda/api/LocationCategoryControllerTestIT.java]: Removed extra test

File: cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java [hunk 5]
```diff
@@ -227,10 +227,9 @@ public void create(@NotNull Context ctx) {
             if (!deserialize.getLocationCategory().getOfficeId().equalsIgnoreCase(CWMS_OFFICE)
                     && (!deserialize.getOfficeId().equalsIgnoreCase(deserialize.getLocationCategory().getOfficeId())
                     || deserialize.getOfficeId().equalsIgnoreCase(CWMS_OFFICE))) {
-                CdaError re = new CdaError("Office ID cannot be CWMS and must match the location category office ID");
-                logger.info(() -> re + System.lineSeparator() + "for request " + ctx.fullUrl());
-                ctx.status(HttpServletResponse.SC_BAD_REQUEST).json(re);
-                return;
+                CdaError re = new CdaError("Location Group office ID cannot be CWMS and must match the "
+                    + "Location Category office ID");
+                throw new IllegalArgumentException(re.toString());
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java]: just pass in the message, no need to create an exception just to call toString() on it

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java [hunk 6]
```diff
@@ -237,10 +258,9 @@ public void create(@NotNull Context ctx) {
             if (!deserialize.getTimeSeriesCategory().getOfficeId().equalsIgnoreCase(CWMS_OFFICE)
                     && (!deserialize.getOfficeId().equalsIgnoreCase(deserialize.getTimeSeriesCategory().getOfficeId())
                     || deserialize.getOfficeId().equalsIgnoreCase(CWMS_OFFICE))) {
-                CdaError re = new CdaError("Office ID cannot be CWMS and must match the TimeSeries category office ID");
-                logger.info(() -> re + System.lineSeparator() + "for request " + ctx.fullUrl());
-                ctx.status(HttpServletResponse.SC_BAD_REQUEST).json(re);
-                return;
+                CdaError re = new CdaError("TimeSeries Group office ID cannot be CWMS and must match the "
+                        + "TimeSeries Category office ID");
+                throw new IllegalArgumentException(re.toString());
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java]: same comment

File: cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java (hunk 5 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/LocationGroupController.java]: Removed CdaError

File: cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java (hunk 6 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesGroupController.java]: Removed CdaError

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**